tempesta_fw/http_parser.c

```diff
@@ -1791,512 +1780,46 @@ __parse_transfer_encoding(TfwHttpMsg *hm, unsigned char *data, size_t len)
-	Resp_HdrExpires,
-	Resp_HdrExpiresV,
-	Resp_HdrK,
-	Resp_HdrKe,
-	Resp_HdrKee,
-	Resp_HdrKeep,
-	Resp_HdrKeep_,
-	Resp_HdrKeep_A,
-	Resp_HdrKeep_Al,
-	Resp_HdrKeep_Ali,
-	Resp_HdrKeep_Aliv,
-	Resp_HdrKeep_Alive,
-	Resp_HdrKeep_AliveV,
-	Resp_HdrL,
-	Resp_HdrLa,
-	Resp_HdrLas,
-	Resp_HdrLast,
-	Resp_HdrLast_,
-	Resp_HdrLast_M,
-	Resp_HdrLast_Mo,
-	Resp_HdrLast_Mod,
-	Resp_HdrLast_Modi,
-	Resp_HdrLast_Modif,
-	Resp_HdrLast_Modifi,
-	Resp_HdrLast_Modifie,
-	Resp_HdrLast_Modified,
-	Resp_HdrLast_ModifiedV,
-	Resp_HdrP,
-	Resp_HdrPr,
-	Resp_HdrPra,
-	Resp_HdrPrag,
-	Resp_HdrPragm,
-	Resp_HdrPragma,
-	Resp_HdrPragmaV,
-	Resp_HdrS,
-	Resp_HdrSe,
-	Resp_HdrSer,
-	Resp_HdrServ,
-	Resp_HdrServe,
-	Resp_HdrServer,
-	Resp_HdrServerV,
-	Resp_HdrT,
-	Resp_HdrTr,
-	Resp_HdrTra,
-	Resp_HdrTran,
-	Resp_HdrTrans,
-	Resp_HdrTransf,
-	Resp_HdrTransfe,
-	Resp_HdrTransfer,
-	Resp_HdrTransfer_,
-	Resp_HdrTransfer_E,
-	Resp_HdrTransfer_En,
-	Resp_HdrTransfer_Enc,
-	Resp_HdrTransfer_Enco,
-	Resp_HdrTransfer_Encod,
-	Resp_HdrTransfer_Encodi,
-	Resp_HdrTransfer_Encodin,
-	Resp_HdrTransfer_Encoding,
-	Resp_HdrTransfer_EncodingV,
-	Resp_HdrDone,
-
-	Resp_BodyUnlimStart,
-	Resp_BodyUnlimRead,
-
-	Resp_StatesNum
-};
-
-/*
- * Helping (interior) FSM states
- * for processing specific parts of an HTTP request.
+/**
+ * Accept header parser, RFC 7321 5.3.2.
  */
-enum {
-	Req_I_0,
-
-	/* Host header */
-	Req_I_H_Start,
-	Req_I_H,
-	Req_I_H_v6,
-	Req_I_H_v6_End,
-	Req_I_H_Port,
-	/* Accept header. */
-	Req_I_Accept,
-	Req_I_AcceptOther,
-	/* Authorization header. */
-	Req_I_Auth,
-	/* Cache-Control header */
-	Req_I_CC,
-	Req_I_CC_m,
-	Req_I_CC_n,
-	Req_I_CC_o,
-	Req_I_CC_MaxAgeV,
-	Req_I_CC_MinFreshV,
-	Req_I_CC_MaxStale,
-	Req_I_CC_MaxStaleV,
-	Req_I_CC_Ext,
-	/* X-Forwarded-For header */
-	Req_I_XFF,
-	Req_I_XFF_Node_Id,
-	Req_I_XFF_Sep,
-	/* User-Agent */
-	Req_I_UserAgent,
-	/* Cookie header */
-	Req_I_CookieStart,
-	Req_I_CookieName,
-	Req_I_CookieVal,
-	Req_I_CookieSemicolon,
-	Req_I_CookieSP,
-	/* Referer header */
-	Req_I_Referer,
-	/* Mark part of URI */
-	Req_I_UriMarkStart,
-	Req_I_UriMarkName,
-	Req_I_UriMarkValue,
-
-	Req_I_EoT,
-};
-
-
 static int
 __req_parse_accept(TfwHttpReq *req, unsigned char *data, size_t len)
 {
 	int r = CSTR_NEQ;
 	__FSM_DECLARE_VARS(req);
 
-	__FSM_START(parser->_i_st) {
+	__FSM_START(parser->_i_st);
 
 	__FSM_STATE(Req_I_Accept) {
-		TRY_STR_LAMBDA("text/html", {
-			__set_bit(TFW_HTTP_B_ACCEPT_HTML, req->flags);
-		}, I_EoT);
-		TRY_STR_LAMBDA("*/*", {
-			__set_bit(TFW_HTTP_B_ACCEPT_HTML, req->flags);
-		}, I_EoT);
+		TRY_STR("text/html", Req_I_Accept, Req_I_AcceptHtml);
+		TRY_STR("*/*", Req_I_Accept, Req_I_AcceptHtml);
 		TRY_STR_INIT();
-		__FSM_I_MOVE_n(Req_I_AcceptOther, 0);
+		__FSM_I_JMP(Req_I_AcceptOther);
+	}
+
+	__FSM_STATE(Req_I_AcceptHtml) {
+		if (IS_WS(c) || c == ',' || c == ';' || IS_CRLF(c)) {
+			__set_bit(TFW_HTTP_B_ACCEPT_HTML, req->flags);
+			__FSM_I_JMP(I_EoT);
+		}
+		__FSM_I_MOVE(Req_I_AcceptOther);
 	}
```

Comment:
It seems that the problem of #1128 is still here: we match `text/htmlbla-bla-bla` in `Req_I_Accept`, pass to `Req_I_AcceptHtml`, then go to the end of block and after `__FSM_I_MOVE(Req_I_AcceptOther)` go to the `Req_I_AcceptOther`, where we try to match `la-bla-bla`.

Reply:
Why's that? We have test case
```C
        FOR_REQ("GET / HTTP/1.1\r\n"
                "Accept: text/htmlK, foo\r\n"                                          
                "Connection: keep-aliveA\r\n"
                "Cache-Control: no-transform\", only-if-cachedd\r\n"                   
                "\r\n");                                                               
        {       
                EXPECT_FALSE(test_bit(TFW_HTTP_B_ACCEPT_HTML, req->flags));            
                EXPECT_FALSE(test_bit(TFW_HTTP_B_CONN_KA, req->flags));
                EXPECT_FALSE(req->cache_ctl.flags & (TFW_HTTP_CC_NO_TRANSFORM          
                                                     | TFW_HTTP_CC_OIFCACHED
                                                     | TFW_HTTP_CC_NO_STORE));         
        
        }
```
What's the problem with falling to `Req_I_AcceptOther` if we do not set `TFW_HTTP_B_ACCEPT_HTML`?

Reply:
It seems that I somewhat incorrect understand the issue: in fact it turns out, that we want to pass such messages, but the appropriate flags just won't be set for them.